[botcom] Share menu

steveruizok opened the pull request:
This PR adds UI for the share menu.

Fun fact: while writing this I found a much better way for us to do our QR codes with SVGs, if we want it!

![localhost_3000_q_f_0 (1)](https://github.com/user-attachments/assets/ed1353c8-f2a9-4e86-992f-eedb3a48827f)

![localhost_3000_q_f_0 (2)](https://github.com/user-attachments/assets/6bc7cf32-5ce7-4b50-addb-65e7669a0a4b)

![localhost_3000_q_f_0](https://github.com/user-attachments/assets/89c434c6-273f-4b14-b778-d6c10bc779ef)

### Change type

- [ ] `bugfix`
- [ ] `improvement`
- [ ] `feature`
- [ ] `api`
- [x] `other`

steveruizok commented:
Some feature differences here:

- projects are either not shared, or shared as readonly or shared as edit. There aren't separate readonly links.
- the "copy link" button also turns on sharing (with an animation that should be noticeable). Beats disabled buttons.

steveruizok commented:
I'm feeling more convinced that we need separate UI for exporting images. It's too confusing, especially around selection, formats, and options like transparent backgrounds.

mimecuvalo commented on apps/dotcom/client/src/tla/utils/qrcode.ts:
```diff
@@ -0,0 +1,12 @@
+export async function createQRCodeImageDataString(url: string) {
+	const QRCode = await import('qrcode')
+	return await QRCode.toString(url, {
```
neat!

mimecuvalo commented on apps/dotcom/client/src/tla/utils/copy.ts:
```diff
@@ -0,0 +1,11 @@
+export async function copyTextToClipboard(text: string) {
```
feels like this would be just useful to expose (either by putting it in packages/utils or from tldraw, either way)

mimecuvalo replied:
(to avoid duplication)

steveruizok replied:
It kind of is a duplication, though the one in the tldraw library takes a blob rather than a string.

mimecuvalo commented:
lemme know if you wanna review this more / or just land and iterate. i think at our current stage of prototyping it's ok to just land since i see the designs are changing pretty fast. up to you!

steveruizok commented:
Ok, I think this is ready for a deeper review now.

- added tabs
- added qr code
- added export tab
- added export preview
- added toggle for inline help

ds300 reviewed (approved):
Love this. So clean and clear.

- the css modules are an improvement
- (nonblocking) the help icon acting as a toggle is weird. It causes some nasty jarring repaints, it doesn't look like a toggle and you can't tell which state it's in without clicking it, when you're on the export tab it doesn't do anything, in general it is not particularly well colocated with the bits of UI it's trying to provide clarity for. We should really adopt tooltips for this kind of thing, it's so much easier to integrate into designs, it's much better UX than this, and it's just a well established pattern that people use every day.

mimecuvalo commented on apps/dotcom/client/src/tla/components/TlaSwitch/switch.module.css:
```diff
@@ -0,0 +1,48 @@
+	transition: all 0.12s;
+}
+
+.switch:after {
```
nit: `::` here and below

mimecuvalo commented on apps/dotcom/client/src/tla/components/TlaFileShareMenu/file-share-menu.module.css:
```diff
@@ -0,0 +1,151 @@
+	border-top: 1px solid var(--tla-color-border);
+}
+
+.controlGroup {
```
can i just say, that one of the things i like about css modules is getting away from the BEM stuff?

steveruizok replied:
Yeah I like it too!

mimecuvalo commented on apps/dotcom/client/src/tla/components/TlaFileShareMenu/TlaFileShareMenuSharePage.tsx:
```diff
@@ -0,0 +1,192 @@
+	const ref = useRef<HTMLImageElement>(null)
+
+	// Save the QR codes in local storage
+	const [qrCode, setQrCode] = useLocalStorageState<string | null>(fileId + 'qr-code-11', null)
```
what is the `code-11` for?

steveruizok replied:
just lazy bumps of the local storage key

mimecuvalo commented on apps/dotcom/client/src/tla/components/TlaFileShareMenu/TlaFileShareMenuExportPage.tsx:
```diff
@@ -0,0 +1,333 @@
+
+	return (
+		<TlaShareMenuControl>
+			<TlaShareMenuControlLabel>Theme...</TlaShareMenuControlLabel>
```
little nit: `…` instead of `...`

mimecuvalo reviewed (approved):
👍 

my only high-level feedback is the `Anyone with the link can...`:
i know we're trying to see if we can make it work so that there's only one link and it's either Edit or View. but seeing it in action on this PR, it _does_ feel like we're losing something, not being able to send an Edit link to some people and a View link to others. feels like we should still consider having two separate links, even if it as clean as single primary CTA.

steveruizok commented on apps/dotcom/client/src/tla/components/TlaFileShareMenu/TlaFileShareMenuExportPage.tsx:
```diff
@@ -0,0 +1,333 @@
+
+	return (
+		<TlaShareMenuControl>
+			<TlaShareMenuControlLabel>Export as...</TlaShareMenuControlLabel>
```
Good catch. I'll drop, made more sense in a previous iteration

steveruizok commented:
> my only high-level feedback is the `Anyone with the link can...`: i know we're trying to see if we can make it work so that there's only one link and it's either Edit or View. but seeing it in action on this PR, it _does_ feel like we're losing something, not being able to send an Edit link to some people and a View link to others. feels like we should still consider having two separate links, even if it as clean as single primary CTA.

Yeah, it does feel like one use case is gone. But I think it's a net improvement on the sort of mental model of how and whether a project is shared. I'd like to revisit this sort of UI again once we have groups and named users, because that could be the place for shared editors + readonly external viewers, or dynamic role switching.

steveruizok commented:
(This access stuff is such a strong conversion moment that I don't think it's been explored much outside of that.)
